[FLINK-3368][Kafka 0.8] Handle leader changes in Kafka Consumer.

rmetzger opened the pull request:
Please see the JIRA for an explanation of the problems.
tl;dr: The Kafka 0.8 consumer now handles broker failures internally, without relying on Flink's checkpointing / job recovery.

The test case which was previously relying on a topology restart is now expecting the consumers to handle the failure.

I also tested this change on a 7 nodes cluster. The job was running for 30 minutes, surviving 4 broker shutdowns.

ndimiduk commented:
I just ran into this as well. Do you mind backporting to 0.10 branch too?

rmetzger commented:
Once this PR has been merged (I expect that some adoptions might still be required) I can try and see if its straightforward to backport. Are you planning to stick to Flink 0.10 for a longer period after 1.0 has been released?
 (I'm asking because so far most users were migrating to the latest release)

ndimiduk commented:
My timeline for adopting Flink 1.0 depends heavily on the amount of code
change required to adopt. I haven't looked closely at my exposure to API
and dependency changes. I also usually avoid jumping directly to production
with Apache X.0.0 community releases, preferring to cut their teeth in a
stage env while community can further stabilize.

On Friday, February 12, 2016, Robert Metzger <notifications@github.com>
wrote:

> Once this PR has been merged (I expect that some adoptions might still be
> required) I can try and see if its straightforward to backport. Are you
> planning to stick to Flink 0.10 for a longer period after 1.0 has been
> released?
> (I'm asking because so far most users were migrating to the latest release)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/flink/pull/1623#issuecomment-183314810>.
>

gyfora commented:
We have tested this and it worked correctly. Should we go ahead a merge it?

rmetzger commented:
Awesome. Thanks a lot for testing it! @StephanEwen wanted to take a look at the PR as well.

StephanEwen commented:
Taking a look right now. Give me a bit to double check...

StephanEwen commented:
I have some comments, here is a first batch:

The `ZooKeeperStringSerializer` seems to be only used in tests, but adds a hard compile-scope dependency to `org.I0Itec:ZkClient`. Moving this to test scope gets rid of this.

There are some old functions that seem to have no more use:

  - KafkaTopicPartitionLeader.replaceIgnoringLeader()
  - KafkaTopicPartition.isContained()


In the `OffsetHandler` / `ZookeeperOffsetHandler`, the `getOffsets(...)` methods can probably get rid of the `Fetcher` (since it does not position the fetcher any more.
  
  
There is a lot of option parsing with no checks and proper exceptions. If someone puts in a wrong value, all they see is an exception in `Integer.valueOf()`. In most other places of the code, there is a lot of effort to give good messages on parse errors.
  
Note: Given all the discussion about dependency clashes / shading and resulting issues, I would like to reduce dependencies as much as possible. The Kafka code adds the Guava dependency simply for the `checkNotNull` method. I would simpyl use `java.util.Object.requiteNotNull` instead and get rid of the dependency.

rmetzger commented:
Thank you for the review. I'll address them soon!

StephanEwen commented:
The `getPartitionsForTopic()` method fails hard immediately, when it does not find the Topic name or partition. There will be no retries. Is that desirable, or can it be that some nodes to not know the topic name, because Kafka's meta data has not properly synced?

rmetzger commented:
So far nobody had this as an issue and during the integration tests and my cluster tests, this issue never occurred. But I agree, there is no good reason to fail hard here. I'll remove the exception and just log the error, so that the retry can happen.

StephanEwen commented:
I think this is in okay shape, quite an improvement over the current version.

There are some things I suggest to re-examine:


Method `commitOffsets()` in "KlinkKafkaConsumer08:397" misses `@Override` annotation.

The "TODO: maybe add a check again here if we are still running." seems worth doing

In many places, you do `if (!closableQueue.addIfOpen(...)) { error }`. Simply use the `add()`´ method, which errors if not open. Also (LegacyFetcher:120): There can never be an IllegalStateException while initially loading in the constructor.

I do not see the use of the list `deadBrokerThreads` in the legacy fetcher. Unless I am overlooking something, I would remove it (simplify the code). Since they will eventually shut down anyways, I think you need not sync on them when leaving the `run()` method of the legacy fetcher.

The check whether a fetcher thread is alive should probably be `thread.getNewPartitionsQueue().isOpen()`, rather than `thread.isAlive()`. That is the flag that is atomically changed and checked with the adding of partitions. 

The fetcher's main thread always blocks 5 seconds before it can notice that the broker threads shut down.
I am wondering if we can make that more snappy, by waking the main thread up when a fetcher thread terminates (by adding a marker element to the queue).


Method `findLeaderForPartitions(...)` in the legacy fetcher also fails hard once it cannot find the leader for a partition (no retries). Is that intended?

The code uses everywhere `Integer.valueOf()`, creating a boxed integer, and then unboxes it. `Integer.parseInt(...)` is the preferrable choice (almost always, I would completely drop using `Integer.valueOf()` in any place).

This loop (in `findLeaderForPartitions(...)`) is either to nifty for me to comprehend, or bogus:

```java
List<KafkaTopicPartitionLeader> topicPartitionWithLeaderList = infoFetcher.getPartitions();
List<FetchPartition> partitionsToAssignInternal = new ArrayList<>(partitionsToAssign);

Map<Node, List<FetchPartition>> leaderToPartitions = new HashMap<>();

for(KafkaTopicPartitionLeader partitionLeader: topicPartitionWithLeaderList) {
	if (partitionsToAssignInternal.size() == 0) {
		// we are done: all partitions are assigned
		break;
	}
	Iterator<FetchPartition> fpIter = partitionsToAssignInternal.iterator();
	while (fpIter.hasNext()) {
		FetchPartition fp = fpIter.next();
		if (fp.topic.equals(partitionLeader.getTopicPartition().getTopic())
				&& fp.partition == partitionLeader.getTopicPartition().getPartition()) {
			
			// we found the leader for one of the fetch partitions
			Node leader = partitionLeader.getLeader();
			List<FetchPartition> partitionsOfLeader = leaderToPartitions.get(leader);
			if (partitionsOfLeader == null) {
				partitionsOfLeader = new ArrayList<>();
				leaderToPartitions.put(leader, partitionsOfLeader);
			}
			partitionsOfLeader.add(fp);
			fpIter.remove();
			break;
		}
	}
}
```
Does this do anything different than the version below? (The internal iteration always finds the
exact same element at the first position and breaks).
```java
List<KafkaTopicPartitionLeader> topicPartitionWithLeaderList = infoFetcher.getPartitions();
Map<Node, List<FetchPartition>> leaderToPartitions = new HashMap<>();

for (KafkaTopicPartitionLeader partitionLeader: topicPartitionWithLeaderList) {
	Node leader = partitionLeader.getLeader();
	List<FetchPartition> partitionsOfLeader = leaderToPartitions.get(leader);
	if (partitionsOfLeader == null) {
		partitionsOfLeader = new ArrayList<>();
		leaderToPartitions.put(leader, partitionsOfLeader);
	}
	partitionsOfLeader.add(fp);
}
```

StephanEwen commented:
I have made a few changes on top of this, which address some of the comments above. Have a look at #1672

rmetzger commented:
Awesome, thank you. I'll take your changes from the pull request and address the remaining issues.

StephanEwen commented:
The comments are mostly simplifications and faster response. I did not find a critical issue, so this should be good to go, actually.

rmetzger commented:
Thank you for the good review. I adopted my IntelliJ inspections to find some of the mentioned issues myself in the future (forgotten @Override, unused methods).

I'll address all your comments and then merge the pull request.

Regarding the `findLeaderForPartitions` method: Its not very well documented and quite hidden in the code, but the `topicPartitionWithLeaderList` list contains the TopicPartitions for ALL requested topics.

From a high level, the method does the following:
- Get a list of FetchPartitions (usually only a few partitions)
- Get the list of topics from the FetchPartitions list and request the partitions for the topics. (Kafka doesn't support getting leaders for a set of partitions)
- Build a Map<Leader, List<FetchPartition>> where only the requested partitions are contained.

I added some comments into the code and renamed some variables to make it more obvious.

StephanEwen commented:
Thanks! And I finally got the meaning of that method with the nested loops ;-)
